### DISCUSSION TITLE
tree view for devtools

### DISCUSSION THREAD

**Discussion by scarf005** [TanStack/query#6035] [Ideas]
![image](https://github.com/TanStack/query/assets/54838975/efcd2fe6-aa85-422d-977d-56b1aa3fefd8)

Hi, i think it would be great to be able to group similar queries in devtools by `queryKey` (and also make it collapsible). for example:

## Before

```yaml
[reviews, 1]
[reviews, 2]
[reviews, 3]
[reviews, search, a]
[reviews, search, 2]
[reviews, search, 3]
```

## After
```yaml
- reviews
  - 1
  - 2
  - 3
  - search
    - "a"
    - 2
    - 3
```

**Comment by TkDodo** [comment]
that would be cool indeed

**Comment by czabaj** [comment]
I have the same idea. Recently, I was adding a [@yornaath/batshit](https://www.npmjs.com/package/@yornaath/batshit) to our stack which allows me to query many requests as if it would be a single GET request, so each are cached independently, but this comes with a price of polluted devtools.

**Comment by dborysov** [comment]
Interesting idea... Though in order to implement it, probably we'll need to sort by query key? 🤔

Or if the queries are sorted by some other term, we could still make a tree view, but without collapsing

**Reply by czabaj** [reply]
Tree-view is grouping, right, so if queries are sorted by e.g. _recently-updated_, the group containing the freshest query should go first and other entries in that group will be sorted by _recently-updated_ as well. Then the second group might contain a fresher query than the last item in the first group, but grouping is more important here than sorting. 

It might be annoying that the groups change position in the list depending on the runtime - the expanded group might be first in the list, then jump to second position, third, then back to first... Sorting by query-key would help here for sure.
